meshmode/array_context.py

```diff
@@ -671,29 +633,35 @@ def wrapper(*args):
     return wrapper
 
 
-def freeze(ary):
+@singledispatch
+def freeze(ary, actx=None):
     r"""Freezes recursively by going through all components of the
     :class:`ArrayContainer` *ary*.
 
     :param ary: a :meth:`~ArrayContext.thaw`\ ed :class:`ArrayContainer`.
     """
-    def _freeze(subary, actx=None):
-        if is_array_container(subary):
-            if actx is None:
-                actx = get_container_context(subary)
-
-            return _map_array_container_with_context(
-                    partial(_freeze, actx=actx),
-                    subary, actx=None, recursive=False)
+    if is_array_container(ary):
+        return _map_array_container(
+                partial(freeze, actx=actx), ary,
+                recursive=False)
+    else:
+        if actx is None:
+            raise TypeError(
+                    f"cannot freeze arrays of type {type(ary).__name__} "
+                    "when actx is not supplied. Try calling actx.freeze "
+                    "directly or supplying an array context")
         else:
-            return actx.freeze(subary)
+            return actx.freeze(ary)
 
-    if not is_array_container(ary):
-        raise TypeError(
-                f"cannot freeze arrays of type {type(ary).__name__}; "
-                "try calling 'ArrayContext.freeze' directly")
 
-    return _freeze(ary)
+@singledispatch
+def _thaw_inner(ary, actx):
```

Comment:
This should be public, right? There may be other things besides `DOFArray` that need to implement their own.

Reply:
It should. At the same time, my goal was to hide it from "end user" types. I'll rename it `thaw_impl` to encourage *cough* use by users.

Reply:
3cc199d